// @spectrum-web-components/help-text/src/HelpTextManager.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { html } from '@spectrum-web-components/base';
import { ifDefined } from '@spectrum-web-components/base/src/directives.js';
export class HelpTextManager {
    constructor(host, { mode } = { mode: 'internal' }) {
        this.hadId = false;
        this.mode = 'internal';
        this.handleSlotchange = ({ target, }) => {
            this.handleHelpText(target);
            this.handleNegativeHelpText(target);
        };
        this.host = host;
        this.instanceCount = HelpTextManager.instanceCount++;
        this.id = `sp-help-text-${this.instanceCount}`;
        this.mode = mode;
    }
    get isInternal() {
        return this.mode === 'internal';
    }
    render(negative) {
        // `pass-through-help-text-${this.instanceCount}` makes the slot effectively unreachable from
        // the outside allowing the `help-text` slot to be preferred while `negative === false`.
        return html `
            <div id=${ifDefined(this.isInternal ? this.id : undefined)}>
                <slot
                    name=${negative
            ? 'negative-help-text'
            : `pass-through-help-text-${this.instanceCount}`}
                    @slotchange=${this.handleSlotchange}
                >
                    <slot name="help-text"></slot>
                </slot>
            </div>
        `;
    }
    addId() {
        const id = this.helpTextElement ? this.helpTextElement.id : this.id;
        const ariaDescribedby = this.host.getAttribute('aria-describedby');
        const descriptors = ariaDescribedby ? ariaDescribedby.split(/\s+/) : [];
        this.hadId = descriptors.indexOf(id) > -1;
        if (this.hadId)
            return;
        descriptors.push(id);
        this.host.setAttribute('aria-describedby', descriptors.join(' '));
        if (this.host.hasAttribute('tabindex')) {
            this.previousTabindex = parseFloat(this.host.getAttribute('tabindex'));
        }
        this.host.tabIndex = 0;
    }
    removeId() {
        const ariaDescribedby = this.host.getAttribute('aria-describedby');
        let descriptors = ariaDescribedby ? ariaDescribedby.split(/\s+/) : [];
        if (!this.hadId) {
            const id = this.helpTextElement ? this.helpTextElement.id : this.id;
            descriptors = descriptors.filter((descriptor) => descriptor !== id);
        }
        if (descriptors.length) {
            this.host.setAttribute('aria-describedby', descriptors.join(' '));
        }
        else {
            this.host.removeAttribute('aria-describedby');
        }
        if (this.helpTextElement)
            return;
        if (this.previousTabindex) {
            this.host.tabIndex = this.previousTabindex;
        }
        else {
            this.host.removeAttribute('tabindex');
        }
    }
    handleHelpText(target) {
        if (this.isInternal)
            return;
        if (this.helpTextElement && this.helpTextElement.id === this.id) {
            this.helpTextElement.removeAttribute('id');
        }
        this.removeId();
        const assignedElements = target.assignedElements();
        const nextHelpTextElement = assignedElements[0];
        this.helpTextElement = nextHelpTextElement;
        if (nextHelpTextElement) {
            if (!nextHelpTextElement.id) {
                nextHelpTextElement.id = this.id;
            }
            this.addId();
        }
    }
    handleNegativeHelpText(target) {
        if (target.name !== 'negative-help-text')
            return;
        const assignedElements = target.assignedElements();
        assignedElements.forEach((el) => (el.variant = 'negative'));
    }
}
HelpTextManager.instanceCount = 0;
