// @spectrum-web-components/button/stories/button-secondary.stories.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { html } from '@spectrum-web-components/base';
import { bellIcon, renderButtonSet } from './index.js';
import '@spectrum-web-components/icons-workflow/icons/sp-icon-help.js';
export default {
    component: 'sp-button',
    title: 'Button/Secondary',
    args: {
        disabled: false,
        quiet: false,
        variant: 'secondary',
    },
    argTypes: {
        disabled: {
            name: 'disabled',
            type: { name: 'boolean', required: false },
            description: 'Disable this control. It will not receive focus or events.',
            table: {
                type: { summary: 'boolean' },
                defaultValue: { summary: false },
            },
            control: {
                type: 'boolean',
            },
        },
        quiet: {
            name: 'quiet',
            type: { name: 'boolean', required: false },
            table: {
                type: { summary: 'boolean' },
                defaultValue: { summary: false },
            },
            control: {
                type: 'boolean',
            },
        },
        variant: {
            name: 'variant',
            type: { name: 'string', required: false },
            description: 'The visual variant to apply to the button.',
            table: {
                type: { summary: 'string' },
                defaultValue: { summary: 'cta' },
            },
            control: {
                type: 'inline-radio',
                options: [
                    'cta',
                    'primary',
                    'secondary',
                    'negative',
                    'overBackground',
                ],
            },
        },
    },
};
const variant = 'secondary';
export const Default = (props) => renderButtonSet(props);
export const quiet = (props) => renderButtonSet(props);
quiet.args = {
    quiet: true,
};
export const withIcon = (props) => {
    return html `
        <style>
            .row {
                padding: 10px;
            }
        </style>
        <div class="row">
            ${renderButtonSet(Object.assign(Object.assign({}, props), { content: html `
                    <sp-icon-help slot="icon"></sp-icon-help>
                    Help
                ` }))}
        </div>
        <div class="row">
            ${renderButtonSet(Object.assign(Object.assign({}, props), { content: html `
                    ${bellIcon} Custom SVG
                ` }))}
        </div>
    `;
};
export const iconSizeOverridden = () => {
    return html `
        <sp-button label="Edit" size="xl" variant=${variant}>
            <sp-icon-help slot="icon" size="s">Testing</sp-icon-help>
        </sp-button>
        <h1>For testing purposes only</h1>
        <p>
            This is a test to ensure that sizing the icon will still work when
            it's in the scope of a parent element. You shouldn't normally do
            this as it deviates from the Spectrum design specification.
        </p>
    `;
};
export const minWidthButton = (props) => {
    return html `
        <style>
            sp-button {
                min-width: 300px;
            }
        </style>
        ${renderButtonSet(props)}
    `;
};
minWidthButton.story = {
    name: 'min-width',
};
const href = 'https://github.com/adobe/spectrum-web-components';
export const link = (props) => renderButtonSet(Object.assign(Object.assign({}, props), { href }));
link.story = {
    name: 'href',
};
export const linkWithTarget = (props) => renderButtonSet(Object.assign(Object.assign({}, props), { href, target: '_blank' }));
linkWithTarget.story = {
    name: 'href with target="_blank"',
};
